### PR TITLE
Dijkstra's algorithm 

### PR DESCRIPTION
**Pull request by Freddievaneijsden**
PR for Issue #32:
This pull request introduces a Dijkstra algorithm implementation for finding the shortest paths in a weighted graph. Key features include:

Initialization: Constructs the algorithm using a WeightedGraph with nodes and edges.

Main Functions:
findShortestPath(WeightedGraph<T> graph, Node<T> start, Node<T> end): Finds the shortest path from start to end.
findAllShortestPaths(WeightedGraph<T> graph, Node<T> start): Finds the shortest paths from start to all nodes.
getDistance(Node<T> target): Return distance from the target node.
getPath(Node<T> target): Returns the path from the start node to the target node.

## Summary by CodeRabbit

- **New Features**
  - Introduced an optimized algorithm for calculating shortest paths in weighted graphs.
  - Added core components to represent and manage graph elements, including nodes, edges, and weighted graphs.

- **Tests**
  - Expanded test coverage for the Dijkstra algorithm to validate distance updates, node visitation, and pathfinding across various scenarios and edge cases.
  - Implemented tests to handle edge cases such as negative weights, self-loops, cyclic graphs, and disconnected graphs.

### CODE REVIEW

**Review by AntonGrubisic — APPROVED** [review]
Looks good after the two minor issues that is now resolved! 
LGTM!

Reviewed by me and @CarlGullberg

**Review by VictoriaNaenfeldtEklund — COMMENTED** [review]
Great job implementing this! I had such a hard time wrapping my head around this algorithm so creeds to you for making it work! 
I have left some comments for you to consider (or ignore if my understanding of the implementation is faulty). Please let me know if you have any questions :)

File: src/main/java/org/fungover/breeze/graph/Dijkstra.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+public class Dijkstra<T> {
+
+    public List<Node<T>> unvisitedNodes = new ArrayList<>();
+    public List<Node<T>> visitedNodes = new ArrayList<>();
+    public WeightedGraph<T> graph;
+
+    public Dijkstra(WeightedGraph<T> graph) {
+        unvisitedNodes.addAll(graph.getNodes());
+        this.graph = graph;
+    }
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/main/java/org/fungover/breeze/graph/Dijkstra.java:14]: Good point, we will fix all!

File: src/main/java/org/fungover/breeze/graph/Dijkstra.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+     * when sum of source node and weight is less than the destinations distance
+     * @param node the current node being processed
+     * @param graph the weighted graph containing the nodes and edges
+     */
+    public void updateDistance(Node<T> node, WeightedGraph<T> graph) {
+        Collection<Edge<T>> edges = graph.getEdges(node);
+        for (Edge<T> edge : edges) {
+            if (edge.getSource().equals(edge.getDestination())) {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (Daymaare) [reply src/main/java/org/fungover/breeze/graph/Dijkstra.java:27]: Thats a valid sugestion! But in this case the edges is suposed to accept self-looping nodes.

File: src/main/java/org/fungover/breeze/graph/Dijkstra.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                break;
+            }
+
+            Optional<Node<T>> optionalNode = findShortestUnvisitedDistance();
+            if (optionalNode.isEmpty()) {
+                break;
+            }
+            currentNode = optionalNode.get();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/main/java/org/fungover/breeze/graph/Dijkstra.java:88]: Added extra flow control to handle disconnected graph + test.

File: src/main/java/org/fungover/breeze/graph/Edge.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    public Edge(Node<T> source, Node<T> destination, double weight) {
+        if (weight < 0.0) {
+            throw new IllegalArgumentException("Weight can't be a negative number");
+        }
+
+        this.source = source;
+        this.destination = destination;
+        this.weight = weight;
+    }
```
[the remark this answers is not in the document]
Comment (Daymaare) [reply src/main/java/org/fungover/breeze/graph/Edge.java:16]: Good point! Null checkers will be added for source and destination. 
Thank you

File: src/main/java/org/fungover/breeze/graph/WeightedGraph.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
+package org.fungover.breeze.graph;
+
+import java.util.*;
+
+public class WeightedGraph<T> implements Graph<T> {
+    public List<Node<T>> listOfNodes = new ArrayList<>();
+    public List<Edge<T>> listOfEdges = new ArrayList<>();
+    public Map<Node<T>, List<Edge<T>>> nodesWithEdges = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/main/java/org/fungover/breeze/graph/WeightedGraph.java:8]: Well spotted, we'll remove the field and it's methods, seems to be code that've been implemented in more efficient ways along the way:)

File: src/main/java/org/fungover/breeze/graph/WeightedGraph.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public Collection<Edge<T>> getEdges(Node<T> node) {
+        List<Edge<T>> edges = new ArrayList<>();
+        for (Edge<T> edge : listOfEdges)
+            if (edge.getSource().equals(node)) {
+                edges.add(edge);
+            }
+        return edges;
+    }
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/main/java/org/fungover/breeze/graph/WeightedGraph.java:42]: The thought initially was that the return type had to correspond to it's interface, but your right, the subtype is also valid and more clean. Thanks!

File: src/main/java/org/fungover/breeze/graph/Dijkstra.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        while (!unvisitedNodes.isEmpty()) {
+            updateDistance(currentNode, graph);
+
+            Optional<Node<T>> optionalNode = findShortestUnvisitedDistance();
+            if (optionalNode.isEmpty()) {
+                break;
+            }
+            currentNode = optionalNode.get();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Daymaare) [reply src/main/java/org/fungover/breeze/graph/Dijkstra.java:126]: Correct! will add that method to handleCyclicGraphs

File: src/main/java/org/fungover/breeze/graph/Dijkstra.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    private void handleCyclicGraphs(WeightedGraph<T> graph) {
+        for (Node<T> node : graph.getNodes()) {
+            node.setDistance(Double.MAX_VALUE);
+            node.setPreviousNode(null);
+        }
+
+        unvisitedNodes.clear();
+        unvisitedNodes.addAll(graph.getNodes());
+    }
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/main/java/org/fungover/breeze/graph/Dijkstra.java:106]: Yes! That should allow Dijkstra's algorithm to run multiple times without interference from previous runs. Will be added.

File: src/test/java/org/fungover/breeze/graph/DijkstraTest.java [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+    @Test
+    @DisplayName("Instantiate edge with negative weight should throw exception")
+    void instantiateEdgeWithNegativeWeightShouldThrowException() {
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
+            Edge<String> edgeWithNegativeWeight = new Edge<>(nodes.get(0), nodes.get(1), -0.1);
+        });
+
+        assertThat(exception.getMessage()).isEqualTo("Weight can't be a negative number");
+    }
```
[the remark this answers is not in the document]
Comment (Daymaare) [reply src/test/java/org/fungover/breeze/graph/DijkstraTest.java:213]: There is only need for one assert, but the local variable a bit unnecessary.

File: src/test/java/org/fungover/breeze/graph/DijkstraTest.java [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        );
+
+        WeightedGraph<String> disconnectedGraph = new WeightedGraph<>(nodes, edges);
+        Dijkstra<String> dijkstra = new Dijkstra<>(disconnectedGraph);
+
+        dijkstra.findShortestPath(disconnectedGraph, nodes.get(0), nodes.get(3));
+
+        assertAll(
+                () -> assertThat(nodes.get(3).getDistance()).isEqualTo(Double.MAX_VALUE),
+                () -> assertThat(dijkstra.getPath(nodes.get(3)).isEmpty())
```
[the remark this answers is not in the document]
Comment (Freddievaneijsden) [reply src/test/java/org/fungover/breeze/graph/DijkstraTest.java:294]: We splited the test into two to check for infinite distance and stop updating nodes if disconnected, hope this works out better!

**Review by AntonGrubisic — APPROVED** [review]
Looks good after the recent changes, ready for merge.

**Review by JesperSandklef — DISMISSED** [review]
Code looks good, it's structured well with good documentation and coverage. Approved

**Review by JesperSandklef — APPROVED** [review]
Changes looks good. Approved

**Review by VictoriaNaenfeldtEklund — APPROVED** [review]
